File: aat-android/src/main/java/ch/bailu/aat/map/layer/gpx/legend/GpxLegendLayer.java
package ch.bailu.aat.map.layer.gpx.legend;

import android.graphics.Color;

import org.mapsforge.core.graphics.Paint;
import org.mapsforge.core.model.Point;

import ch.bailu.aat_lib.map.MapContext;
import ch.bailu.aat_lib.map.MapMetrics;
import ch.bailu.aat.map.MapPaint;
import ch.bailu.aat.map.layer.gpx.GpxLayer;
import ch.bailu.aat_lib.preferences.StorageInterface;

public final class GpxLegendLayer extends GpxLayer {
    private final LegendWalker walker;

    private Paint backgroundPaint, framePaint;

    private int color = Color.WHITE;

    public GpxLegendLayer(LegendWalker w) {
        walker=w;

        initPaint();
    }

    @Override
    public void drawInside(MapContext mcontext) {
        if (color != getColor()) {
            color = getColor();

            initPaint(mcontext.getMetrics());
        }

        walker.init(mcontext, backgroundPaint, framePaint);
        walker.walkTrack(getGpxList());
    }



    private void initPaint(MapMetrics metrics) {
        backgroundPaint = MapPaint.createBackgroundPaint(color);
        framePaint = MapPaint.createEdgePaintLine(metrics.getDensity());
        framePaint.setColor(color);
    }


    private void initPaint() {
        backgroundPaint = MapPaint.createBackgroundPaint(color);
        framePaint = MapPaint.createBackgroundPaint(color);
    }


    @Override
    public boolean onTap(Point tapXY) {
        return false;
    }

    @Override
    public void onPreferencesChanged(StorageInterface s, String key) {}


    @Override
    public void onAttached() {

    }

    @Override
    public void onDetached() {

    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/map/layer/gpx/legend/GpxLegendLayer.kt
package ch.bailu.aat_lib.map.layer.gpx.legend

import ch.bailu.aat_lib.lib.color.ARGB
import ch.bailu.aat_lib.map.MapContext
import ch.bailu.aat_lib.map.MapMetrics
import ch.bailu.aat_lib.map.MapPaint.createBackgroundPaint
import ch.bailu.aat_lib.map.MapPaint.createEdgePaintLine
import ch.bailu.aat_lib.map.layer.gpx.GpxLayer
import org.mapsforge.core.graphics.Paint

class GpxLegendLayer(private val walker: LegendWalker) : GpxLayer() {
    private var backgroundPaint = createBackgroundPaint()
    private var framePaint: Paint = createBackgroundPaint()
    private var color = ARGB.WHITE

    init {
        initPaint()
    }

    override fun drawInside(mcontext: MapContext) {
        if (color != colorFromIID) {
            color = colorFromIID
            initPaint(mcontext.getMetrics())
        }
        walker.init(mcontext, backgroundPaint, framePaint)
        walker.walkTrack(gpxList)
    }

    private fun initPaint(metrics: MapMetrics) {
        backgroundPaint = createBackgroundPaint(color)
        framePaint = createEdgePaintLine(metrics.getDensity())
        framePaint.color = color
    }

    private fun initPaint() {
        backgroundPaint = createBackgroundPaint(color)
        framePaint = createBackgroundPaint(color)
    }
}
